lib/util/findGraphRoots.js

```diff
@@ -0,0 +1,229 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author Tobias Koppers @sokra
+*/
+
+"use strict";
+
+const NO_MARKER = 0;
+const IN_PROGRESS_MARKER = 1;
+const DONE_MARKER = 2;
+const DONE_MAYBE_ROOT_CYCLE_MARKER = 3;
+const DONE_AND_ROOT_MARKER = 4;
+
+class Node {
+	constructor(item) {
+		this.item = item;
+		this.dependencies = new Set();
+		this.marker = NO_MARKER;
+		this.cycle = undefined;
+		this.incoming = 0;
+	}
+}
+
+class Cycle {
+	constructor() {
+		this.nodes = new Set();
+	}
+}
+
+/**
+ * @typedef {Object} StackEntry
+ * @property {Node} node
+ * @property {Node[]} openEdges
+ */
+
+/**
+ * @template T
+ * @param {Iterable<T>} items list of items
+ * @param {function(T): Iterable<T>} getDependencies function to get dependencies of an item (items that are not in list are ignored)
+ * @returns {Iterable<T>} graph roots of the items
+ */
+module.exports = (items, getDependencies) => {
+	const itemToNode = new Map();
+	for (const item of items) {
+		const node = new Node(item);
+		itemToNode.set(item, node);
+	}
+
+	// early exit when there is only a single item
+	if (itemToNode.size <= 1) return items;
```

Comment:
Can we check it in first line of function?
```js
if (items.length <= 1) return items
// or // if (items.size <= 1) return items
```

Reply:
It won't catch duplicates

Reply:
and `items` is an `Iterable` which has no `size` or `length`